// @bitcoinerlab/descriptors/dist/applyPR2137.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.applyPR2137 = void 0;
const utils_1 = require("bip174/src/lib/utils");
const bip341_1 = require("bitcoinjs-lib/src/payments/bip341");
const bip371_1 = require("bitcoinjs-lib/src/psbt/bip371");
const crypto_1 = require("bitcoinjs-lib/src/crypto");
const toXOnly = (pubKey) => pubKey.length === 32 ? pubKey : pubKey.slice(1, 33);
function range(n) {
    return [...Array(n).keys()];
}
function tapBranchHash(a, b) {
    return (0, crypto_1.taggedHash)('TapBranch', Buffer.concat([a, b]));
}
function calculateScriptTreeMerkleRoot(leafHashes) {
    if (!leafHashes || leafHashes.length === 0) {
        return undefined;
    }
    // sort the leaf nodes
    leafHashes.sort(Buffer.compare);
    // create the initial hash node
    let currentLevel = leafHashes;
    // build Merkle Tree
    while (currentLevel.length > 1) {
        const nextLevel = [];
        for (let i = 0; i < currentLevel.length; i += 2) {
            const left = currentLevel[i];
            const right = i + 1 < currentLevel.length ? currentLevel[i + 1] : left;
            nextLevel.push(i + 1 < currentLevel.length ? tapBranchHash(left, right) : left);
        }
        currentLevel = nextLevel;
    }
    return currentLevel[0];
}
function getTweakSignersFromHD(inputIndex, inputs, hdKeyPair) {
    const input = (0, utils_1.checkForInput)(inputs, inputIndex);
    if (!input.tapBip32Derivation || input.tapBip32Derivation.length === 0) {
        throw new Error('Need tapBip32Derivation to sign with HD');
    }
    const myDerivations = input.tapBip32Derivation
        .map(bipDv => {
        if (bipDv.masterFingerprint.equals(hdKeyPair.fingerprint)) {
            return bipDv;
        }
        else {
            return;
        }
    })
        .filter(v => !!v);
    if (myDerivations.length === 0) {
        throw new Error('Need one tapBip32Derivation masterFingerprint to match the HDSigner fingerprint');
    }
    const signers = myDerivations.map(bipDv => {
        const node = hdKeyPair.derivePath(bipDv.path);
        if (!bipDv.pubkey.equals(toXOnly(node.publicKey))) {
            throw new Error('pubkey did not match tapBip32Derivation');
        }
        const h = calculateScriptTreeMerkleRoot(bipDv.leafHashes);
        const tweakValue = (0, bip341_1.tapTweakHash)(toXOnly(node.publicKey), h);
        return node.tweak(tweakValue);
    });
    return signers;
}
function getSignersFromHD(inputIndex, inputs, hdKeyPair) {
    const input = (0, utils_1.checkForInput)(inputs, inputIndex);
    if ((0, bip371_1.isTaprootInput)(input)) {
        return getTweakSignersFromHD(inputIndex, inputs, hdKeyPair);
    }
    if (!input.bip32Derivation || input.bip32Derivation.length === 0) {
        throw new Error('Need bip32Derivation to sign with HD');
    }
    const myDerivations = input.bip32Derivation
        .map(bipDv => {
        if (bipDv.masterFingerprint.equals(hdKeyPair.fingerprint)) {
            return bipDv;
        }
        else {
            return;
        }
    })
        .filter(v => !!v);
    if (myDerivations.length === 0) {
        throw new Error('Need one bip32Derivation masterFingerprint to match the HDSigner fingerprint');
    }
    const signers = myDerivations.map(bipDv => {
        const node = hdKeyPair.derivePath(bipDv.path);
        if (!bipDv.pubkey.equals(node.publicKey)) {
            throw new Error('pubkey did not match bip32Derivation');
        }
        return node;
    });
    return signers;
}
const applyPR2137 = (psbt) => {
    psbt.signInputHD = function signInputHD(inputIndex, hdKeyPair, sighashTypes) {
        if (!hdKeyPair || !hdKeyPair.publicKey || !hdKeyPair.fingerprint) {
            throw new Error('Need HDSigner to sign input');
        }
        const signers = getSignersFromHD(inputIndex, this.data.inputs, hdKeyPair);
        signers.forEach(signer => this.signInput(inputIndex, signer, sighashTypes));
        return this;
    };
    psbt.signAllInputsHD = function signAllInputsHD(hdKeyPair, sighashTypes) {
        if (!hdKeyPair || !hdKeyPair.publicKey || !hdKeyPair.fingerprint) {
            throw new Error('Need HDSigner to sign input');
        }
        const results = [];
        for (const i of range(psbt.data.inputs.length)) {
            try {
                psbt.signInputHD(i, hdKeyPair, sighashTypes);
                results.push(true);
            }
            catch (err) {
                results.push(false);
            }
        }
        if (results.every(v => v === false)) {
            throw new Error('No inputs were signed');
        }
        return psbt;
    };
};
exports.applyPR2137 = applyPR2137;
