// @bitcoinerlab/coinselect/dist/index.js
"use strict";
// Copyright (c) 2023 Jose-Luis Landabaso - https://bitcoinerlab.com
// Distributed under the MIT software license
Object.defineProperty(exports, "__esModule", { value: true });
exports.avoidChange = exports.addUntilReach = exports.maxFunds = exports.dustThreshold = exports.vsize = exports.coinselect = exports.MAX_FEE_RATE = exports.DUST_RELAY_FEE_RATE = void 0;
// Same as in Bitcoin Core
exports.DUST_RELAY_FEE_RATE = 3;
//10 times larger than 2017 peak
exports.MAX_FEE_RATE = 10000;
var coinselect_1 = require("./coinselect");
Object.defineProperty(exports, "coinselect", { enumerable: true, get: function () { return coinselect_1.coinselect; } });
var vsize_1 = require("./vsize");
Object.defineProperty(exports, "vsize", { enumerable: true, get: function () { return vsize_1.vsize; } });
var dust_1 = require("./dust");
Object.defineProperty(exports, "dustThreshold", { enumerable: true, get: function () { return dust_1.dustThreshold; } });
var maxFunds_1 = require("./algos/maxFunds");
Object.defineProperty(exports, "maxFunds", { enumerable: true, get: function () { return maxFunds_1.maxFunds; } });
var addUntilReach_1 = require("./algos/addUntilReach");
Object.defineProperty(exports, "addUntilReach", { enumerable: true, get: function () { return addUntilReach_1.addUntilReach; } });
var avoidChange_1 = require("./algos/avoidChange");
Object.defineProperty(exports, "avoidChange", { enumerable: true, get: function () { return avoidChange_1.avoidChange; } });
